hphp/runtime/vm/jit/vasm-simplify.cpp

```diff
@@ -473,6 +473,29 @@ bool simplify(Env& env, const pop& inst, Vlabel b, size_t i) {
   });
 }
 
+/*
+ * Eliminate masking values with -1 in andXi.
+ * andbi s0=0xff s1=val0 d=val1 -> movb s=val0 d=val1
+ * andli s0=0xffffffff s1=val0 d=val1 -> movl s=val0 d=val1
+ */
+template<typename mov, typename andi>
+bool simplify_andi(Env& env, const andi& inst, Vlabel b, size_t i) {
+  if (inst.s0.l() != -1) return false;
+  return simplify_impl(env, b, i, [&] (Vout& v) {
+    if (inst.s1 != inst.d)
+      v << mov{inst.s1, inst.d};
+    return 1;
+  });
+}
```

Comment:
This transformation is unsound.  You also need to ensure that nobody is using the SF output operand of the `and*{}` instruction.  You can look around this file for similar checks.